### PR TITLE
Use process dictionary instead of ets

### LINKED ISSUE #17: Consider not using ets for each connection
Currently, one ets table is used per connection. This means that if there are more than ~1400 connections, then the ets table limit will be hit.

The table is created at https://github.com/Aircloak/phoenix_gen_socket_client/blob/4fbddbf2b4f94e81131bf84ed2a6c31988f7fcac/lib/gen_socket_client.ex#L231

I have a quick fix at the moment using a named table (on this branch https://github.com/VoiceLayer/phoenix_gen_socket_client/tree/fix/ets-limit), however I am sure there are much better solutions we could implement.

### PR DESCRIPTION
**Pull request by Gazler**
With ets, when there were many connections, it was possible to hit the
ets table limit after ~1400 connections. Using the process dictionary
instead maintains the same functionality and API, whilst fixing the
ets table limit issue.

This closes #17

### CODE REVIEW

File: lib/gen_socket_client.ex [hunk 1]
```diff
@@ -169,11 +168,11 @@ defmodule Phoenix.Channels.GenSocketClient do
   @spec push(transport, topic, event, out_payload) :: {:ok, ref} | {:error, reason::any}
   def push(%{transport_pid: nil}, _topic, _event, _payload), do: {:error, :disconnected}
   def push(transport, topic, event, payload) do
-    ref = next_ref(topic, transport.message_refs)
+    ref = next_ref(topic)
     cond do
       # first message on a channel must always be a join
       event != "phx_join" and ref == 1 ->
-        :ets.delete(transport.message_refs, topic)
+        :erlang.erase(topic)
```
Comment (sasa1977) [inline lib/gen_socket_client.ex]: [Process.delete/1](https://hexdocs.pm/elixir/Process.html#delete/1)?

File: lib/gen_socket_client.ex [hunk 2]
```diff
@@ -321,16 +319,24 @@ defmodule Phoenix.Channels.GenSocketClient do
...
   end
 
   defp transport(state),
-    do: Map.take(state, [:transport_mod, :transport_pid, :message_refs, :serializer])
-
-  defp next_ref(topic, message_refs),
-    do: :ets.update_counter(message_refs, topic, 1, {topic, 0})
+    do: Map.take(state, [:transport_mod, :transport_pid, :serializer])
+
+  defp next_ref(topic) do
```
Comment (sasa1977) [inline lib/gen_socket_client.ex]: Same thing in this function. There's `Process.get` and `Process.put` for that. And using `Process.get/2`, you could remove the `case` here.

File: lib/gen_socket_client.ex [hunk 3]
```diff
@@ -321,16 +319,24 @@ defmodule Phoenix.Channels.GenSocketClient do
...
+    case :erlang.get(topic) do
```
Comment (sasa1977) [inline lib/gen_socket_client.ex]: Also, I think we should tag the procdict key, so it doesn't get mixed up with something by the user. For example, `{__MODULE__, topic}`, or something like that.

File: lib/gen_socket_client.ex [hunk 4]
```diff
@@ -321,16 +319,24 @@ defmodule Phoenix.Channels.GenSocketClient do
   end
 
   defp reinit(state) do
-    :ets.delete_all_objects(state.message_refs)
+    :erlang.erase()
```
Comment (Gazler) [inline lib/gen_socket_client.ex]: @sasa1977 What should we do here? We shouldn't risk deleting user keys.

Comment (Gazler) [inline lib/gen_socket_client.ex]: Maybe we could get all the keys and delete where the first argument is the module name?

Comment (Gazler) [inline lib/gen_socket_client.ex]: Excuse the monologue, I have gone with:

```elixir
    Process.get_keys()
    |> Enum.filter(&match?({__MODULE__, _}, &1))
    |> Enum.map(&Process.delete/1)
```

Comment (sasa1977) [inline lib/gen_socket_client.ex]: Good catch, we definitely shouldn't delete all other keys (which might include an RNG state for example). Your proposal sounds sensible. You can use`Process.get_keys/0` for that.

**Comment by Gazler** [comment]
@sasa1977 Given that diff is so small, I have force pushed on the existing commit instead of creating a new one for review. It should have all your changes.

I meant to flag the `:erlang.erase()` think earlier. Also TIL that there are wrappers for the process dictionary in Elixir.

I can confirm that it does work with > 1400 connections. I've tested with 5000 and it works perfectly.

**Comment by sasa1977** [comment]
@Gazler Version 1.1.1 containing your changes is now published on hex.pm. Thanks for contributing!

**Pull request merged**